### PR TITLE
coprocessor: add a util method 'in_union' for RpnFnCallExtra/ScalarFunc

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

## What have you changed? (mandatory)

Add a utility method `in_union` for `RpnFnCallExtra/ScalarFunc`.

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- Misc (other changes)

## How has this PR been tested? (mandatory)

unit tests

### CODE REVIEW

File: src/coprocessor/dag/expr/mod.rs [hunk 1]
```diff
@@ -66,6 +66,19 @@ pub struct ScalarFunc {
...
+impl ScalarFunc {
+    /// Indicates whether the current expression is evaluated in union statement
+    pub fn in_union(&self) -> bool {
+        if self.implicit_args.is_empty() {
+            return false;
+        }
+        match self.implicit_args[0] {
+            Datum::I64(1) => true,
+            _ => false,
+        }
```
Comment (kennytm) [inline src/coprocessor/dag/expr/mod.rs]: The whole thing could simply be

```rust
    self.implicit_args.get(0) == Some(&Datum::I64(1))
```

File: src/coprocessor/dag/rpn_expr/types/function.rs [hunk 2]
```diff
@@ -136,6 +136,23 @@ pub struct RpnFnCallExtra<'a> {
...
+    /// but some CAST expressions inserted by TiKV coprocessor use an empty vector to represent
+    /// the `inUnion` flag is false.
+    pub fn in_union(&self) -> bool {
+        if self.implicit_args.is_empty() {
+            return false;
+        }
+        match self.implicit_args[0] {
+            ScalarValue::Int(Some(1)) => true,
+            _ => false,
+        }
```
Comment (kennytm) [inline src/coprocessor/dag/rpn_expr/types/function.rs]: ditto

File: src/coprocessor/dag/rpn_expr/types/function.rs [hunk 3]
```diff
@@ -334,3 +345,127 @@ thread_local! {
...
+                    let call_extra = RpnFnCallExtra {
+                        ret_field_type: field_type,
+                        implicit_args,
+                    };
+                    assert_eq!(
+                        call_extra.in_union(),
+                        expect,
+                        "expect: {}, got: {}",
+                        expect,
+                        !expect
```
Comment (kennytm) [inline src/coprocessor/dag/rpn_expr/types/function.rs]: The extra comment looks unnecessary.

File: src/coprocessor/dag/expr/mod.rs [hunk 4]
```diff
@@ -66,6 +66,13 @@ pub struct ScalarFunc {
     cus_rng: CusRng,
 }
 
+impl ScalarFunc {
+    /// Indicates whether the current expression is evaluated in union statement
+    pub fn in_union(&self) -> bool {
+        self.implicit_args.get(0) == Some(&Datum::I64(1))
```
Comment (kennytm) [inline src/coprocessor/dag/expr/mod.rs]: Is there any doc mentioning the first argument being "1" means "in union"? If yes please add a link.

**Comment by breeswish** [comment]
I don't think it's appropriate to add a util function like this. `in_union` is a UDF dependent argument, instead of a common argument for all UDFs. Thus, it is not suitable to add a util function checking `in_union` for all UDFs. i.e. `implicit_args[0] == 1` does not imply `in_union`.

**Comment by lonng** [comment]
@breeswish How about use an individual util function?

**Comment by breeswish** [comment]
@lonng LGTM. It should be a cast specific function.

**Comment by lonng** [comment]
@breeswish The parameter type is different, I think it's OK to place in individual cast module (We will remove the traditional expression eventually).

File: src/coprocessor/dag/rpn_expr/impl_cast.rs [hunk 5]
```diff
@@ -164,3 +177,25 @@ fn cast_json_as_real(ctx: &mut EvalContext, val: &Option<Json>) -> Result<Option
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    #[allow(clippy::iter_skip_next)]
```
Comment (lonng) [inline src/coprocessor/dag/rpn_expr/impl_cast.rs]:
```suggestion
```

**Pull request merged**